### ISSUE TITLE
Style/RedundantSelfAssignment

### ISSUE THREAD

**Issue by coding-red-panda** [rubocop/rubocop#13750]
Take the following class as an example:

```ruby
class TestObject
  def self.data
    Thread.current[:data] || []
  end

  def self.data=(value)
    Thread.current[:data] = value
  end

  def self.test(*values)
     self.data = data.concat(values)
  end
end
```

In the above example, the cop will flag the method `self.test` for violating the self-assignment using `.concat`.
The cop will recommend in this case that the method can be written as just `data.concat(values)`.
And in theory, it's correct if the following change is made:

```ruby
  def self.data
    Thread.current[:data] ||= []
  end
```

In the end, I removed all usages of `.concat` in this class and rely on the `+` operator, taking the performance hit.
The question I would like to raise, is whether the cop is able to detect that using `.concat` in this situation would break the functionality, since the `self.data` method always returns a new empty array initially, and needs to re-assign the value to store it (unless the `||=` operator is used).

--------

## Expected behavior

Method should not be flagged in this particular sitution.

## Actual behavior

The method gets flagged as violation.

## Steps to reproduce the problem

See the code example

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
1.70.0 (using Parser 3.3.7.0, rubocop-ast 1.37.0, analyzing as Ruby 3.3, running on ruby 3.3.5) [arm64-darwin24]
  - rubocop-factory_bot 2.26.1
  - rubocop-graphql 1.5.4
  - rubocop-performance 1.23.1
  - rubocop-rails 2.29.0
  - rubocop-rspec 3.3.0
  - rubocop-rspec_rails 2.30.0
```

## Extra

Was asked on the Discord to file an issue for this.
I am totally fine with having this closed as non-issue if the core team decides as such.
It just sparked an interesting discussion on the Disord ;)

**Comment by koic** [comment]
Details are described in #13750, avoiding `self.foo =` is likely better approach. Thank you for the feedback.
